diff -Nur a/xbmc/utils/EGLUtils.cpp c/xbmc/utils/EGLUtils.cpp
--- a/xbmc/utils/EGLUtils.cpp	2018-04-01 00:19:02.000000000 +0200
+++ c/xbmc/utils/EGLUtils.cpp	2018-04-10 22:26:02.554269479 +0200
@@ -103,7 +103,7 @@
     EGL_NONE
   };
 
-#if defined(EGL_EXT_platform_base) && defined(EGL_KHR_platform_gbm) && defined(HAVE_GBM)
+#if 0 //defined(EGL_EXT_platform_base) && defined(EGL_KHR_platform_gbm) && defined(HAVE_GBM)
   if (m_eglDisplay == EGL_NO_DISPLAY &&
       CEGLUtils::HasExtension(EGL_NO_DISPLAY, "EGL_EXT_platform_base") &&
       CEGLUtils::HasExtension(EGL_NO_DISPLAY, "EGL_KHR_platform_gbm"))
diff -Nur a/xbmc/utils/EGLUtils.h c/xbmc/utils/EGLUtils.h
--- a/xbmc/utils/EGLUtils.h	2018-04-01 00:19:02.000000000 +0200
+++ c/xbmc/utils/EGLUtils.h	2018-04-10 21:13:26.872801875 +0200
@@ -23,7 +23,8 @@
 #include <string>
 #include <stdexcept>
 
-#include <EGL/egl.h>
+#include <FBDEV/eglplatform.h>
+#include <FBDEV/egl.h>
 
 #include "StringUtils.h"
 
diff -Nur a/xbmc/windowing/gbm/GBMUtils.cpp c/xbmc/windowing/gbm/GBMUtils.cpp
--- a/xbmc/windowing/gbm/GBMUtils.cpp	2018-04-01 00:19:02.000000000 +0200
+++ c/xbmc/windowing/gbm/GBMUtils.cpp	2018-04-10 21:10:07.783124383 +0200
@@ -69,6 +69,14 @@
                                                                          width,
                                                                          height);
 
+  if (!m_nativeWindow)
+  {
+    fbdev_window *nativeWindow = new fbdev_window;
+    nativeWindow->width = width;
+    nativeWindow->height = height;
+    m_nativeWindow = static_cast<EGLNativeWindowType>(nativeWindow);
+  }
+
   return true;
 }
 
@@ -84,6 +92,11 @@
     gbm_surface_destroy(m_surface);
     m_surface = nullptr;
   }
+
+  if(m_nativeWindow)
+  {
+    m_nativeWindow = static_cast<EGLNativeWindowType>(NULL);
+  }
 }
 
 struct gbm_bo *CGBMUtils::LockFrontBuffer()
diff -Nur a/xbmc/windowing/gbm/GBMUtils.h c/xbmc/windowing/gbm/GBMUtils.h
--- a/xbmc/windowing/gbm/GBMUtils.h	2018-04-01 00:19:02.000000000 +0200
+++ c/xbmc/windowing/gbm/GBMUtils.h	2018-04-10 21:11:57.103055493 +0200
@@ -21,6 +21,8 @@
 #pragma once
 
 #include <gbm.h>
+#include <FBDEV/eglplatform.h>
+#include <FBDEV/egl.h>
 
 class CGBMUtils
 {
@@ -36,6 +38,7 @@
 
   struct gbm_device *m_device = nullptr;
   struct gbm_surface *m_surface = nullptr;
+  EGLNativeWindowType m_nativeWindow = nullptr;
 
 protected:
   struct gbm_bo *m_bo = nullptr;
diff -Nur a/xbmc/windowing/gbm/WinSystemGbm.cpp c/xbmc/windowing/gbm/WinSystemGbm.cpp
--- a/xbmc/windowing/gbm/WinSystemGbm.cpp	2018-04-06 16:46:57.000000000 +0200
+++ c/xbmc/windowing/gbm/WinSystemGbm.cpp	2018-04-10 22:57:10.103956748 +0200
@@ -197,7 +197,8 @@
 {
   struct gbm_bo *bo = m_GBM->LockFrontBuffer();
 
-  m_DRM->FlipPage(bo, rendered, videoLayer);
+  if (videoLayer)
+	m_DRM->FlipPage(bo, rendered, videoLayer);
 
   m_GBM->ReleaseBuffer();
 }
diff -Nur a/xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp c/xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp
--- a/xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp	2018-04-01 00:19:02.000000000 +0200
+++ c/xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp	2018-04-10 21:12:12.547346460 +0200
@@ -50,7 +50,7 @@
     return false;
   }
 
-  if (!m_pGLContext.CreateDisplay(m_GBM->m_device,
+  if (!m_pGLContext.CreateDisplay(0,
                                   EGL_OPENGL_ES2_BIT,
                                   EGL_OPENGL_ES_API))
   {
@@ -99,7 +99,7 @@
     return false;
   }
 
-  if (!m_pGLContext.CreateSurface(reinterpret_cast<EGLNativeWindowType>(m_GBM->m_surface)))
+  if (!m_pGLContext.CreateSurface(reinterpret_cast<EGLNativeWindowType>(m_GBM->m_nativeWindow)))
   {
     return false;
   }
