diff --git a/drivers/gpu/drm/drm_fb_helper.c b/drivers/gpu/drm/drm_fb_helper.c
index 035784ddd133..54e93a97b6fd 100644
--- a/drivers/gpu/drm/drm_fb_helper.c
+++ b/drivers/gpu/drm/drm_fb_helper.c
@@ -1595,9 +1595,6 @@ int drm_fb_helper_check_var(struct fb_var_screeninfo *var,
 	case 16:
 		depth = (var->green.length == 6) ? 16 : 15;
 		break;
-	case 32:
-		depth = (var->transp.length > 0) ? 32 : 24;
-		break;
 	default:
 		depth = var->bits_per_pixel;
 		break;
